### PR TITLE
Add async WebSocket PubsubClient

### LINKED ISSUE #22006: Does pubsub_client have async impl ?
#### Problem

Hey Guys, I am using solana pubsub client as part of my program. I found that WS is using tungstenite-rs but without having async impl (tokio-tungstenite). The performance of the pubsub client is blocked by the receiving loop. Morever, the websocket spec doesn't have the ping/pong functionality to have heartbeat check. 

#### Proposed Solution
1. Should I implement the async version of the pubsub client and making the pr in the future or do we have another plan ? 
2. Do we have alternative method to make the heartbeat check ? (such as subscribe to slotNotification)

discussion welcome

### PR DESCRIPTION
**Pull request by fanatid**
Closes #22006

### CODE REVIEW

**Comment by mvines** [comment]
There's also an effort to provide an async rpc client at https://github.com/solana-labs/solana/pull/20683
It'd be great if both these new clients felt the same from an API perspective

File: client/src/pubsub_client_async.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        },
+        MaybeTlsStream, WebSocketStream,
+    },
+    url::Url,
+};
+
+pub type PubsubClientResult<T = ()> = Result<T, PubsubClientError>;
+
+#[derive(Debug, Error)]
+pub enum PubsubClientError {
```
Comment (mvines) [inline client/src/pubsub_client_async.rs:45]: Can this `PubsubClientError` be shared with the existing enum in `pubsub_client`?

**Comment by mvines** [comment]
Hey @fanatid, I wasn't able to push to this PR so I created a new PR at https://github.com/solana-labs/solana/pull/22767 that builds on your work here.  Mind taking a look at that PR to see if it looks good to you? If so I think we can close this PR and land the async PubsubClient over their

**Comment by fanatid** [comment]
@mvines thank you that updated PR! I did not have time for this in the last few weeks :disappointed: 
Hope that next step will be upgrading sync PubsubClient to using async :slightly_smiling_face:

**Pull request closed without merging**